src/GraphQL/Instrumentation/GraphQLTelemetryProvider.cs

```diff
@@ -0,0 +1,149 @@
+#if NET5_0_OR_GREATER
+
+using System.Diagnostics;
+using GraphQL.DI;
+using GraphQL.Execution;
+using GraphQL.Types;
+using GraphQL.Validation;
+using GraphQLParser.AST;
+
+namespace GraphQL.Instrumentation;
+
+/// <summary>
+/// Provides telemetry through the <see cref="Activity">System.Diagnostics.Activity API</see> as an <see cref="IConfigureExecution"/> implementation.
+/// Derive from this class to add additional telemetry.
+/// <para>
+/// To use a derived class, call <see cref="GraphQLBuilderExtensions.ConfigureExecution{TConfigureExecution}(IGraphQLBuilder)"/> with the type of your derived class.
+/// </para>
+/// </summary>
+public class GraphQLTelemetryProvider : IConfigureExecution
+{
+    private readonly GraphQLTelemetryOptions _telemetryOptions;
+    private const string ACTIVITY_OPERATION_NAME = "graphql";
+
+    /// <summary>
+    /// Returns an <see cref="System.Diagnostics.ActivitySource"/> instance to be used for GraphQL.NET telemetry.
+    /// </summary>
+    protected static ActivitySource ActivitySource { get; }
+
+    /// <summary>
+    /// Gets the source name used for telemetry.
+    /// </summary>
+    public static string SourceName => "GraphQL";
+
+    static GraphQLTelemetryProvider()
+    {
+        ActivitySource = new ActivitySource(SourceName, typeof(GraphQLTelemetryProvider).Assembly.GetNuGetVersion());
+    }
+
+    /// <summary>
+    /// Initializes a new instance with the specified options.
+    /// </summary>
+    public GraphQLTelemetryProvider(GraphQLTelemetryOptions options)
+    {
+        _telemetryOptions = options;
+    }
+
+    /// <inheritdoc/>
+    public virtual float SortOrder => GraphQLBuilderExtensions.SORT_ORDER_CONFIGURATION;
+
+    /// <inheritdoc/>
+    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        // start the Activity, in fact Activity.Stop() will be called from within Activity.Dispose() at the end of using block 
+        using var activity = await StartActivityAsync(options).ConfigureAwait(false);
+
+        // do not record any telemetry if there are no listeners or it decided not to sample the current request
+        if (activity == null)
+            return await next(options).ConfigureAwait(false);
+
+        // record the requested operation name and optionally the GraphQL document
+        await SetInitialTagsAsync(activity, options).ConfigureAwait(false);
+
+        // record the operation type and the operation name (which may be specified within the
+        // document even if not specified in the request)
+        options.Listeners.Add(new TelemetryListener(this, activity, options));
```

Comment:
It would be nice to somehow eliminate the need to allocate `TelemetryListener` to store contextual info for setting tags some time later. @Kielek @gao-artur By the way this (additional allocations etc) is the reason why I would keep `UseTelemetry` opt-in. I don't want to sacrifice the overall performance in favor of convenience of use for some users. I understand that this is just one more allocation (for now but things may change) but please understand the general idea - I'm guided by the "pay for play" principle.

Reply:
BTW another performance hit - instantiating async/await state machines, even when no activities tracked.

Reply:
I disagree.  The "performance hit" is per-request, and having a few allocations per-request is not worth mentioning.  There is also no "instantiating async/await state machines" within the document listener at all, whether it is enabled or not.  There is only a single async/await state machine per request, and that is on the call to `ExecuteAsync`.  This is not worth mentioning, and could be optimized away when there are no listeners (at the cost of some ugly code).

Reply:
Opt-in design of new features is harmless. Opt-out design **may be** harmful.

Reply:
Ok, then let's move on since we appear to be in agreement that opt-in is the right approach for now.

Reply:
👍